### PR TITLE
Repeated fields in Dart generated validation

### PR DESCRIPTION
**Pull request by dmdashenkov**
In this PR we add support for generating validation Dart code for `repeated` and `map` fields.

Those fields can be checked according to the `(required)` and `(distinct)` options.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@dmitrykuzmin, PTAL.

File: tools/dart-code-gen/lib/src/field_validator_factory.dart [hunk 1]
```diff
@@ -47,6 +47,66 @@ class FieldValidatorFactory {
...
+    Expression _requiredMissing() {
+        return violationRef.call([literalString('Field must be set.'),
+                                  literalString(validatorFactory.fullTypeName),
+                                  literalList([field.name])]);
+    }
+}
+
+/// A [FieldValidatorFactory] for non-`repeated` fields.
+///
+class ScalarFieldValidatorFactory extends FieldValidatorFactory {
```
Comment (dmitrykuzmin) [inline tools/dart-code-gen/lib/src/field_validator_factory.dart]: I think the name is a bit misleading as this factory validates both [scalar](https://developers.google.com/protocol-buffers/docs/proto#scalar) and message types.

File: tools/smoke-tests/dart-codegen-tests/test/validation_test.dart [hunk 2]
```diff
@@ -57,17 +58,61 @@ void main() {
...
+            test('empty required repeated messages', () {
+                var name = PersonName()..givenName = 'William';
+                var contact = Contact()
+                    ..name = name
+                    ..category = Contact_Category.WORK
+                    ..email.addAll([EmailAddress()..value = 'will@example.com', EmailAddress()]);
+                checkMissing(contact, 'email');
+            });
+
+            test('missing requied repeated numbers', () {
```
Comment (dmitrykuzmin) [inline tools/smoke-tests/dart-codegen-tests/test/validation_test.dart]:
```suggestion
            test('missing required repeated numbers', () {
```

File: tools/smoke-tests/dart-codegen-tests/test/validation_test.dart [hunk 3]
```diff
@@ -57,17 +58,61 @@ void main() {
...
+                var ticket = LotteryTicket()
+                    ..magicNumber = 42;
+                checkMissing(ticket, 'numbers');
+            });
+
+            test('missing requied map values', () {
```
Comment (dmitrykuzmin) [inline tools/smoke-tests/dart-codegen-tests/test/validation_test.dart]:
```suggestion
            test('missing required map values', () {
```

**Pull request merged**